### PR TITLE
Implement selecting maintenance windows in Struts actions

### PR DESCRIPTION
**Pull request by hustodemon**
## What does this PR change?

Implement the maintenance window picker in Struts:
- backend support: implement listing upcoming maintenance windows
- implement the component in JSP
- populate the component based on upcoming maint. windows for selected system(s) and action
- read the selected maint. window date time when user schedules an action

In order to implement that, the current date / maintenance window picker needed to be extended to accept more options (`maintenance_window`). After that, all places, where we populate the date picker and where we read from it were visited and enhanced with maint. window support.

Review commit-by-commit.

TODO:
- [x] jUnit tests
- [x] todos in the code
- [x] corner cases:
  - [x] handle no maint. windows selected scenario (need to prevent the scheduling)
  - [x] reload the action chains editor on change
- [x] more manual testing
- [ ] docs

## GUI diff

Before:
ss-mw-datepick

After:
Select maint. windows:
ss-mw-pick

Maint. windows selector for action chain:
ss-mw-ac-

Maint. windows picker hidden since systems in ssm have multiple schedules assigned:
ss-mw-multi

- [x] **DONE**

## Documentation
updated https://github.com/SUSE/spacewalk/issues/11254

- [x] **DONE**

## Test coverage
- Unit tests were added

- [x] **DONE**

## Links

Tracks https://github.com/SUSE/spacewalk/issues/11230

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/maintenance/MaintenanceManager.java [hunk 1]
```diff
@@ -96,6 +106,101 @@ public static MaintenanceManager instance() {
...
+        }
+
+        // there are multiple schedules, we warn the user about it
+        if (schedules.size() > 1) {
+            throw new IllegalStateException("Multiple schedules");
+        }
+
+        PeriodList periods = schedules.iterator().next().getCalendarOpt()
+                .flatMap(c -> parseCalendar(c))
+                .map(c -> calculateUpcomingPeriodList(c, empty(), new Date(), 10))
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:145]: `empty()` is not correct here, right? You need to check if schedules has a multi calendar and if yes, pass the schedule name to it.

Comment (hustodemon) [reply java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:145]: Good catch! This overflowed here from the prototype somehow :)

File: java/code/src/com/redhat/rhn/frontend/action/configuration/sdc/FileListConfirmSetupAction.java [hunk 2]
```diff
@@ -30,13 +31,14 @@
...
 import java.util.List;
+import java.util.Set;
 
 
 /**
  * FileListConfirmSetupAction, for sdc configuration
  * @version $Rev$
  */
-public class FileListConfirmSetupAction extends BaseListAction {
+public abstract class FileListConfirmSetupAction extends BaseListAction implements MaintenanceWindowsAware {
```
Comment (hustodemon) [inline java/code/src/com/redhat/rhn/frontend/action/configuration/sdc/FileListConfirmSetupAction.java:41]: Note for the reviewer: I made the original class abstract and split the actual implementation into 2 concrete classes (above) since struts action using this class resulted in 2 different backend actions (diff or deploy).

File: java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml [hunk 3]
```diff
@@ -6157,6 +6157,15 @@ organization. You may grant or remove the organization administrator role in the
...
       </trans-unit>
+      <trans-unit id="schedule.jsp.maint-window">
+        <source>Maintenance window:</source>
+      </trans-unit>
+      <trans-unit id="schedule.jsp.multiple_maintenance_schedules">
+        <source>There are multiple maintenance schedules for selected systems. Make sure that systems in the set use at most 1 maintenance schedule if you want to schedule by date or maintenance window.</source>
+      </trans-unit>
+      <trans-unit id="schedule.jsp.no_maintenance_windows">
+        <source>No upcoming maintenance windows</source>
+      </trans-unit>
```
Comment (mcalmer) [inline java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:6168]: They need to be copied over to all other language files.
But we can also decide to do it later on the maintenance-windows branch

Comment (hustodemon) [reply java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:6168]: Right, I noticed nobody does this, so I also decided not to :wink:

File: java/code/src/com/redhat/rhn/common/localization/LocalizationService.java [hunk 4]
```diff
@@ -393,6 +394,17 @@ public String formatDate(Date date) {
...
+    /**
+     * Format the instant and let the service determine the locale
+     * @param instant the instant
+     * @return String representation of given date.
+     */
+    public String formatDate(Instant instant) {
+        Context ctx = Context.getCurrentContext();
+        Date date = Date.from(instant);
+        return formatDate(date, ctx.getLocale());
+    }
```
[the remark this answers is not in the document]
Comment (hustodemon) [reply java/code/src/com/redhat/rhn/common/localization/LocalizationService.java:406]: Hm, that looks better, thanks!

**Comment by hustodemon** [comment]
Thank you for your reviews :+1:  Given the fact that my computer works now, I'll address the corner cases now and will add them to this PR. These shouldn't (hopefully) be big.

**Comment by hustodemon** [comment]
@mcalmer @parlt91 ok, I think the PR is now finally complete. It'd be great if at least one of you checked the newest commits (starting from "Use review suggestion for formatting instant"). They are really small and they have meaningful descriptions.

**Comment by mcalmer** [comment]
LGTM

**Comment by hustodemon** [comment]
Merging, thank you!
